kong/db/strategies/postgres/connector.lua

```diff
@@ -996,24 +1037,36 @@ local _M = {}
 
 function _M.new(kong_config)
   local config = {
-    host       = kong_config.pg_host,
-    port       = kong_config.pg_port,
-    timeout    = kong_config.pg_timeout,
-    user       = kong_config.pg_user,
-    password   = kong_config.pg_password,
-    database   = kong_config.pg_database,
-    schema     = kong_config.pg_schema or "",
-    ssl        = kong_config.pg_ssl,
-    ssl_verify = kong_config.pg_ssl_verify,
-    cafile     = kong_config.lua_ssl_trusted_certificate,
+    host        = kong_config.pg_host,
+    port        = kong_config.pg_port,
+    timeout     = kong_config.pg_timeout,
+    user        = kong_config.pg_user,
+    password    = kong_config.pg_password,
+    database    = kong_config.pg_database,
+    schema      = kong_config.pg_schema or "",
+    ssl         = kong_config.pg_ssl,
+    ssl_verify  = kong_config.pg_ssl_verify,
+    cafile      = kong_config.lua_ssl_trusted_certificate,
+    sem_max     = kong_config.pg_semaphore_max or 0,
+    sem_timeout = kong_config.pg_semaphore_timeout or 60,
   }
 
   local db = pgmoon.new(config)
 
+  local sem
+  if config.sem_max > 0 then
+    local err
+    sem, err = semaphore.new(config.sem_max)
+    if not sem then
+      error("Failure creating Postgres connector semaphore: " .. err)
```

Comment:
Note: all error messages in Kong are supposed to be lower-case. I also suggest that we do not prevent initialization in case the semaphore allocation fails, but that we instead issue a `CRIT` error message. The reason being that in case of a change in `ngx.semaphore` semantics in the future, or other unexpected failure scenarios, we avoid a situation in which many users are unable to even _start_ their Kong instances. Since a lack of semaphore will not prevent Kong from running, but will prevent it from running in _optimal_, _predictable_ conditions, a `CRIT` log is appropriate.

Note: other instantiation failures such as `kong.dao` or `kong.cache` should ideally make Kong's initialization fail, but currently do not, alas, due to the worker-events library only being instantiated in `init_worker`. Something that has been communicated about and is awaiting for a fix to initialize worker-events in `init`. Currently, such failures to initialize singletons are logged a `CRIT`, but really should be logged with `EMERG`.